### ISSUE TITLE
01-logistic-regression-bank-marketing-campaign-data.ipynb

### ISSUE THREAD

**Issue by mezcolantriz** [4GeeksAcademy/alemantellini-machine-learning#2]
@alemantellini 

### 1. **EDA Excepcional** 

**El EDA es de calidad profesional. Demuestra excelente comprensión de los datos, de verdad que enhorabuena.**

### 2. **Selección Inteligente de Features**
-  Decisión razonada de eliminar `duration` (data leakage) - aunque podría haber otra donde es importante, en algunas ocasiones puede ayudar a mejorar el modelo. Ya que es influyente si miras las correlaciones. Pero tu decisión está bien planteada y es válida.
- Justificación clara de `pdays` (99% del valor es 999 = sin contacto previo)
- Eliminación de indicadores redundantes (`nr.employed`)
- Mantenimiento de variables potencialmente informativas

**Demuestras entendimiento del problema empresarial, no solo técnico.**

---

## **PROBLEMAS - ORDEN DE OPERACIONES**

Tienes **algunos fallos comunes de inicio en este punto**, te voy contando:

### **PROBLEMA 1: Data Leakage en One-Hot Encoding**

Tu código actual:
```python
# INCORRECTO
total_data = pd.get_dummies(total_data, drop_first=True)  # Encoding en TODO el dataset

mi_x = total_data.drop("y", axis=1)
mi_y = total_data["y"]

X_train, X_test, y_train, y_test = train_test_split(mi_x, mi_y, test_size=0.2)
```

**¿Por qué es problema?**
- `pd.get_dummies()` analiza **todas las categorías** en el dataset completo
- Luego haces split, pero ya las categorías están "contaminadas" con información de test
- Las métricas reportadas (accuracy 0.897) son **infladas**
- En producción, si una categoría nueva aparece, el modelo fallará

**Orden correcto:**
```python
# CORRECTO
X = total_data.drop("y", axis=1)
y = total_data["y"]

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
# Split PRIMERO

# One-Hot Encoding DESPUÉS
from sklearn.preprocessing import OneHotEncoder
categorical_cols = ['job', 'marital', 'education', 'default', 'housing', 'loan', 'contact', 'poutcome']
numeric_cols = ['age', 'campaign', 'previous', 'emp.var.rate', 'cons.price.idx', 'cons.conf.idx', 'euribor3m']

encoder = OneHotEncoder(sparse_output=False, handle_unknown='ignore')

# Fit SOLO en training data
X_train_encoded = encoder.fit_transform(X_train[categorical_cols])
X_test_encoded = encoder.transform(X_test[categorical_cols])  # transform test sin leakage

# Combinar con numéricas escaladas
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
X_train_numeric = scaler.fit_transform(X_train[numeric_cols])
X_test_numeric = scaler.transform(X_test[numeric_cols])

# Concatenar
X_train_final = np.hstack([X_train_numeric, X_train_encoded])
X_test_final = np.hstack([X_test_numeric, X_test_encoded])
```

**Impacto:** Tus métricas son optimistas. El verdadero rendimiento en test sería menor.

---

### **PROBLEMA 2: Pipeline de Optimización Incompleto**

Mira tu código de optimización:

```python
# Paso 1: Entrenas modelo SIN escalado
logmodel = LogisticRegression(max_iter=1000)
logmodel.fit(X_train, y_train)
y_pred = logmodel.predict(X_test)
# Reporta: accuracy 0.897

# Paso 2: AHORA haces scaling (¡después de ya tener predicciones!)
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)  # Modifica los datos
X_test = scaler.transform(X_test)

# Paso 3: Creas NUEVO modelo pero NUNCA lo entrenas
logmodel = LogisticRegression(C=0.5)  # ← Solo declarado, no fit()

# Paso 4: Reportas métricas del modelo ANTERIOR (sin escalado)
print(classification_report(y_test, y_pred))  # Sigue siendo del primer modelo
```

**Problema estructura:**
- El modelo optimizado (C=0.5) nunca fue entrenado con datos escalados
- El reporte que muestras es del modelo sin scaling
- No hay comparación real de si C=0.5 es mejor o peor
- El scaling modificó X_train/X_test pero son usadas nunca en el modelo optimizado

Scaling en si no es una optimización, es parte fundamental de nuestra regresión.

---

### **PROBLEMA 3: No Consideras Desbalance de Clases**

En tu EDA viste:
- **No**: ~88% de casos
- **Yes**: ~12% de casos

Un modelo que predice "no" para todo logra 88% accuracy. Pero:
```python
# Deberías usar
LogisticRegression(max_iter=1000, class_weight='balanced')
```

Esto penaliza errores en la clase minoritaria (yes), mejorando recall. Digamos que iguala para que les de la misma importancia a pesar de su desbalance.

### **PROBLEMA 4: Boxplots de Variables Categóricas**

```python
# Incorrecto
sns.boxplot(ax=axs[0, 0], data=total_data, y="job")
```

Boxplot es para variables numéricas. Para categóricas mejor usa:
```python
# Correcto
sns.countplot(data=total_data, x="job", hue="y")
```

### **PROBLEMA 5: Cálculo de Statstics Duplicado**

Varios cálculos de IQR son exactamente iguales. Podrías reutilizar código:
```python
def detect_outliers_iqr(data, column_name):
    Q1 = data[column_name].quantile(0.25)
    Q3 = data[column_name].quantile(0.75)
    IQR = Q3 - Q1
    
    lower = Q1 - 1.5 * IQR
    upper = Q3 + 1.5 * IQR
    
    outliers = data[(data[column_name] < lower) | (data[column_name] > upper)]
    return outliers, lower, upper

# Usar para cada variable
for col in numeric_cols:
    outliers, lower, upper = detect_outliers_iqr(total_data, col)
    print(f"{col}: {len(outliers)} outliers")
```

---

## **RESUMEN DE CAMBIOS NECESARIOS**

**CRÍTICO (Corrección Obligatoria):**
1. Mover `train_test_split` ANTES de `pd.get_dummies` y... (paso 2)
2. Cambiar a `OneHotEncod` con `fit_transform` / `transform`
3. Reentrenar modelo DESPUÉS de scaling
4. Comparar modelos

**IMPORTANTE:**
5. Agregar `class_weight='balanced'` para manejar desbalance
6. Usar ROC-AUC, Precision, Recall (no solo Accuracy)
7. Podrías cambiar boxplots de categóricas (usar countplots)

**OPCIONAL:**
8. Refactorizar código de IQR para evitar repetición

Tu EDA es profesional. Entiendes los datos.

El pipeline de ML tiene errores que inflan tus métricas (los ERRORES COMUNES a estas alturas). Las decisiones sobre timing (cuándo hacer qué operación) son fundamentales en ML.

Un EDA perfecto + modelo defectuoso = resultados no confiables. 
Un EDA bueno + modelo robusto = solución confiable.

De verdad que enhorabuena por tu trabajo, vamos a dejarlo perfect :)
